presto-orc/src/main/java/com/facebook/presto/orc/OrcWriteValidation.java

```diff
@@ -236,35 +255,146 @@ public void validateRowGroupStatistics(OrcDataSourceId orcDataSourceId, long str
+        // aggregate such statistics before the validation
+        Map<Integer, List<ColumnStatistics>> actualColumnStatisticsByColumn = new HashMap<>();
+        for (Entry<StreamId, List<RowGroupIndex>> entry : actualRowGroupStatistics.entrySet()) {
+            int column = entry.getKey().getColumn();
+            ColumnStatistics actual = entry.getValue().get(rowGroupIndex).getColumnStatistics();
+            List<ColumnStatistics> aggregateStats = actualColumnStatisticsByColumn.computeIfAbsent(column, (key) -> new ArrayList<>());
+            aggregateStats.add(actual);
+
+            // Regular nodes have only 1 ColumnStatistics in the ROW_INDEX, flattened nodes
+            // might have zero or more column statistics.
+            if (aggregateStats.size() != 1 && !flattenedValueNodes.contains(column)) {
+                throw new OrcCorruptionException(
+                        orcDataSourceId,
+                        "Unexpected multiple column statistics for node %s in row group %s in stripe at offset %s",
+                        column,
+                        rowGroupIndex,
+                        stripeOffset);
+            }
+        }
+
+        return actualColumnStatisticsByColumn.entrySet().stream()
+                .collect(Collectors.toMap(Entry::getKey, entry -> mergeColumnStatistics(entry.getValue())));
+    }
+
+    private Map<Integer, Integer> getFlattenedKeyToMapNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
+    {
+        ImmutableMap.Builder<Integer, Integer> keyNodeToMapNode = ImmutableMap.builder();
+        flattenedNodes.forEach(mapNodeIndex -> keyNodeToMapNode.put(orcTypes.get(mapNodeIndex).getFieldTypeIndex(0), mapNodeIndex));
+        return keyNodeToMapNode.build();
+    }
+
+    private Map<Integer, Set<Integer>> getFlattenedMapToValueNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
+    {
+        ImmutableMap.Builder<Integer, Set<Integer>> keyNodeToMapNode = ImmutableMap.builder();
+        flattenedNodes.forEach(mapNodeIndex -> keyNodeToMapNode.put(mapNodeIndex, getFlattenedValueNodes(ImmutableSet.of(mapNodeIndex), orcTypes)));
+        return keyNodeToMapNode.build();
+    }
+
+    private Set<Integer> getFlattenedValueNodes(Set<Integer> flattenedNodes, List<OrcType> orcTypes)
+    {
+        ImmutableSet.Builder<Integer> valueNodes = ImmutableSet.builder();
+        List<Integer> stack = new ArrayList<>();
+
+        for (Integer mapNodeIndex : flattenedNodes) {
+            int valueNodeIndex = orcTypes.get(mapNodeIndex).getFieldTypeIndex(1); // get map.value node index
+            stack.add(valueNodeIndex);
+
+            while (!stack.isEmpty()) {
+                int nodeIndex = stack.remove(stack.size() - 1);
+                valueNodes.add(nodeIndex);
+                OrcType orcType = orcTypes.get(nodeIndex);
+                stack.addAll(orcType.getFieldTypeIndexes());
+            }
+        }
```

Comment:
This seems like we are setting the value of the each node to all the other flat map nodes. 

There are files, where multiple flat map columns are present and this sounds confusing to me.

Reply:
Correct. This is exactly the intention here - collect all flat map value nodes present in the file.